### PR TITLE
Core/BattlePets: Basics for Battle Pets

### PR DESCRIPTION
**Pull request by horn**
Basically this enables the whole pet journal including learning pets, putting them in cages, unlocking the battle slots and putting the pets into them, releasing, renaming pets and setting them as favorite.

There are still some things to do, read the comments in the code.
Maybe some server-side checks should be added even when the client does them too.
Also there are probably more places where to send error packets.
<b>I give no warranty for world sql, some data may be incorrect or missing.</b> (I've tried to do my best.)

Known issues: Summoning battle pet is incomplete (some visual stuff and dismissing), I'll fix that in the next PR.

Any feedback would be great.

### CODE REVIEW

File: src/server/game/BattlePets/BattlePetMgr.cpp [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+    for (auto itr : sBattlePetSpeciesStateStore)
+        _battlePetSpeciesStates[itr->SpeciesID][BattlePetState(itr->State)] = itr->Value;
+
+    LoadAvailablePetBreeds();
+    LoadDefaultPetQualities();
+}
+
+void BattlePetMgr::LoadAvailablePetBreeds()
+{
+    QueryResult result = WorldDatabase.Query("SELECT speciesId, breedId FROM battle_pet_breeds");
```
Comment (MitchesD) [inline src/server/game/BattlePets/BattlePetMgr.cpp]: maybe use prepared statements?

Comment (Shauren) [inline src/server/game/BattlePets/BattlePetMgr.cpp]: No, this only runs once

File: src/server/game/BattlePets/BattlePetMgr.cpp [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+void BattlePetMgr::LoadFromDB(PreparedQueryResult pets, PreparedQueryResult slots)
+{
+    if (pets)
+    {
+        do
+        {
+            Field* fields = pets->Fetch();
+            if (BattlePetSpeciesEntry const* speciesEntry = sBattlePetSpeciesStore.LookupEntry(fields[1].GetUInt32()))
+            {
+                uint32 species = fields[1].GetUInt32();
```
Comment (DDuarte) [inline src/server/game/BattlePets/BattlePetMgr.cpp]: This is getting the same field as the line above. Perhaps move this variable outside of the if?

File: src/server/game/BattlePets/BattlePetMgr.cpp [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+                ++itr;
+                break;
+        }
+    }
+
+    stmt = LoginDatabase.GetPreparedStatement(LOGIN_DEL_BATTLE_PET_SLOTS);
+    stmt->setUInt32(0, _owner->GetBattlenetAccountId());
+    trans->Append(stmt);
+
+    for (WorldPackets::BattlePet::BattlePetSlot slot : _slots)
```
Comment (DDuarte) [inline src/server/game/BattlePets/BattlePetMgr.cpp]: a const& could be used here

File: src/server/game/BattlePets/BattlePetMgr.cpp [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+        stmt = LoginDatabase.GetPreparedStatement(LOGIN_INS_BATTLE_PET_SLOTS);
+        stmt->setUInt8(0, slot.Index);
+        stmt->setUInt32(1, _owner->GetBattlenetAccountId());
+        stmt->setUInt64(2, slot.Pet.Guid.GetCounter());
+        stmt->setBool(3, slot.Locked);
+        trans->Append(stmt);
+    }
+}
+
+BattlePetMgr::BattlePet* BattlePetMgr::GetPet(ObjectGuid guid)
```
Comment (DDuarte) [inline src/server/game/BattlePets/BattlePetMgr.cpp]: this method can be made const, I guess

File: src/server/game/BattlePets/BattlePetMgr.cpp [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+    _slots[slot].Locked = false;
+
+    WorldPackets::BattlePet::PetBattleSlotUpdates updates;
+    updates.Slots.push_back(_slots[slot]);
+    updates.AutoSlotted = false; // what's this?
+    updates.NewSlot = true; // causes the "new slot unlocked" bubble to appear
+    _owner->SendPacket(updates.Write());
+}
+
+std::vector<BattlePetMgr::BattlePet> BattlePetMgr::GetLearnedPets()
```
Comment (DDuarte) [inline src/server/game/BattlePets/BattlePetMgr.cpp]: perhaps make this method const

File: src/server/game/BattlePets/BattlePetMgr.h [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    void LoadFromDB(PreparedQueryResult pets, PreparedQueryResult slots);
+    void SaveToDB(SQLTransaction& trans);
+
+    BattlePet* GetPet(ObjectGuid guid);
+    void AddPet(uint32 species, uint32 creatureId, uint16 breed, uint8 quality, uint16 level = 1);
+    void RemovePet(ObjectGuid guid);
+
+    uint8 GetPetCount(uint32 species) const;
+
+    WorldPackets::BattlePet::BattlePetSlot* GetSlot(uint8 slot) { return &_slots[slot]; }
```
Comment (DDuarte) [inline src/server/game/BattlePets/BattlePetMgr.h]: const method

File: src/server/game/Server/Packets/BattlePetPackets.cpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "BattlePetPackets.h"
+
+ByteBuffer& operator<<(ByteBuffer& data, WorldPackets::BattlePet::BattlePetSlot const& slot)
+{
+    data << (slot.Pet.Guid.IsEmpty() ? ObjectGuid::Create<HighGuid::BattlePet>(0) : slot.Pet.Guid);
+    data << uint32(0); // Collar ID - what does it mean?
```
Comment (DDuarte) [inline src/server/game/Server/Packets/BattlePetPackets.cpp]: shouldn't this be in the struct BattlePetSlot (even if we don't know yet what it does)?

File: src/server/game/Server/Packets/BattlePetPackets.cpp [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+void WorldPackets::BattlePet::BattlePetModifyName::Read()
+{
+    _worldPacket >> PetGuid;
+    uint32 nameLength = _worldPacket.ReadBits(7);
+    bool hasDeclinedNames = _worldPacket.ReadBit();
+    Name = _worldPacket.ReadString(nameLength);
+
+    if (hasDeclinedNames)
+    {
+        uint8 declinedNameLengths[5];
```
Comment (DDuarte) [inline src/server/game/Server/Packets/BattlePetPackets.cpp]: Perhaps you can use this that already is in Unit.h:

```c++
#define MAX_DECLINED_NAME_CASES 5

struct DeclinedName
{
    std::string name[MAX_DECLINED_NAME_CASES];
};
```

File: src/server/game/Server/Packets/BattlePetPackets.h [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+        class BattlePetSetFlags final : public ClientPacket
+        {
+        public:
+            BattlePetSetFlags(WorldPacket&& packet) : ClientPacket(CMSG_BATTLE_PET_SET_FLAGS, std::move(packet)) { }
+
+            void Read() override;
+
+            ObjectGuid PetGuid;
+            uint32 Flags;
```
Comment (DDuarte) [inline src/server/game/Server/Packets/BattlePetPackets.h]: Missing initialization

File: src/server/game/Server/Packets/BattlePetPackets.h [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+        };
+
+        class BattlePetError final : public ServerPacket
+        {
+        public:
+            BattlePetError() : ServerPacket(SMSG_BATTLE_PET_ERROR) { }
+
+            WorldPacket const* Write() override;
+
+            uint8 Result;
```
Comment (DDuarte) [inline src/server/game/Server/Packets/BattlePetPackets.h]: Missing initialization

File: sql/updates/world/battle_pets_6x_world.sql [hunk 11]
```diff
@@ -0,0 +1,3921 @@
...
+(1644,3),
+(1645,3),
+(1646,3),
+(1647,3),
+(1648,3),
+(1649,3),
+(1651,3),
+(1652,3),
+(1653,3),
+(1654,3)
```
Comment (Ladel) [inline sql/updates/world/battle_pets_6x_world.sql]: missing comma

File: src/server/game/Server/Packets/BattlePetPackets.h [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+
+            WorldPacket const* Write() override;
+
+            ObjectGuid PetGuid;
+        };
+
+        class BattlePetError final : public ServerPacket
+        {
+        public:
+            BattlePetError() : ServerPacket(SMSG_BATTLE_PET_ERROR) { }
```
Comment (DDuarte) [inline src/server/game/Server/Packets/BattlePetPackets.h]: Give an initial size to this packet: `ServerPacket(SMSG_BATTLE_PET_ERROR, 5) { }`

File: src/server/game/Server/WorldSession.cpp [hunk 13]
```diff
@@ -132,7 +133,8 @@ WorldSession::WorldSession(uint32 id, std::string&& name, uint32 battlenetAccoun
     _RBACData(NULL),
     expireTime(60000), // 1 min after socket loss, session is deleted
     forceExit(false),
-    m_currentBankerGUID()
+    m_currentBankerGUID(),
+    _battlePetMgr(Trinity::make_unique<BattlePetMgr>(this))
```
Comment (DDuarte) [inline src/server/game/Server/WorldSession.cpp]: Shouldn't this be tied to a Player and not to WorldSession?

File: src/server/game/Server/WorldSession.h [hunk 14]
```diff
@@ -940,6 +952,9 @@ class WorldSession
         uint32 GetRecruiterId() const { return recruiterId; }
         bool IsARecruiter() const { return isRecruiter; }
 
+        // Battle Pets
+        BattlePetMgr* GetBattlePetMgr() { return _battlePetMgr.get(); }
```
Comment (DDuarte) [inline src/server/game/Server/WorldSession.h]: Make const (but perhaps move to player, as I questioned above)

File: sql/updates/characters/battle_pets_6x_characters_item_instance.sql [hunk 15]
```diff
@@ -0,0 +1,5 @@
+ALTER TABLE `item_instance`
```
Comment (DDuarte) [inline sql/updates/characters/battle_pets_6x_characters_item_instance.sql]: Missing changes in `characters_database.sql`

File: src/server/game/Server/WorldSession.cpp (hunk 13 above)
Comment (horn) [inline src/server/game/Server/WorldSession.cpp]: Nope, we've discussed it with Shauren and made decision to put it here. It's account wide, there's no better place.

File: src/server/game/Server/Packets/BattlePetPackets.cpp (hunk 8 above)
Comment (horn) [inline src/server/game/Server/Packets/BattlePetPackets.cpp]: These are only lengths, not the names. For strings I use variable <code>DeclinedName Declined;</code>

Comment (DDuarte) [inline src/server/game/Server/Packets/BattlePetPackets.cpp]: Yarr, just replace 5 with MAX_DECLINED_NAME_CASES

File: src/server/game/Server/Packets/BattlePetPackets.cpp [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+    data << pet.Quality;
+    data.WriteBits(pet.Name.size(), 7);
+    data.WriteBit(!pet.Owner.IsEmpty()); // HasOwnerInfo
+    data.WriteBit(pet.Name.empty()); // NoRename
+    data.FlushBits();
+
+    if (!pet.Owner.IsEmpty())
+    {
+        data << pet.Owner;
+        data << pet.Owner.GetRealmId(); // Virtual
```
Comment (Warpten) [inline src/server/game/Server/Packets/BattlePetPackets.cpp]: ehm?

**Comment by Warpten** [comment]
```c++
DWORD __thiscall sub_D3B4CE(int this, int a2, int a3)
{
  int v3; // esi@1
  DWORD result; // eax@1
  DWORD v5; // [sp+4h] [bp-4h]@1

  v3 = this;
  sub_9C3405(a2, this + 24);
  v5 = 0;
  sub_410BE7(a2, &v5);
  result = v5;
  *(v3 + 16) = v5;
  return result;
}
```

(20253)

This should be ClientBattlePetCageDateError but i dont see it in your diff...

File: src/server/game/BattlePets/BattlePetMgr.h [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+#include "BattlePetPackets.h"
+
+#define MAX_PET_BATTLE_SLOTS 3
+#define MAX_BATTLE_PETS_PER_SPECIES 3
+#define BATTLE_PET_CAGE_ITEM_ID 82800
+
+// TODO: fix values in this enum
+enum BattlePetError
+{
+    BATTLEPETRESULT_CANT_HAVE_MORE_PETS_OF_THAT_TYPE = 8,
```
Comment (Warpten) [inline src/server/game/BattlePets/BattlePetMgr.h]: ClientBattlePetError
3 = ERR_CANT_HAVE_MORE_PETS
7 = ERR_PET_TOO_HIGH_LEVEL_TO_UNCAGE
8 = ERR_CANT_HAVE_MORE_PETS_OF_THAT_TYPE

other opcode (811C19 on 20253, i dont have any names, probably queue feedback)
1 = ERR_PETBATTLE_QUEUE_QUEUED
4 = ERR_PETBATTLE_QUEUE_JOIN_FAILED
6 = ERR_PETBATTLE_QUEUE_JOURNAL_LOCK
7 = ERR_PETBATTLE_QUEUE_NOT_WHILE_NEUTRAL
5 = depends on another byte sent
* 1 = ERR_PETBATTLE_QUEUE_SLOT_LOCKED
* 2 = ERR_PETBATTLE_QUEUE_SLOT_EMPTY
* 3 = ERR_PETBATTLE_QUEUE_SLOT_NO_TRACKER
* 4 = ERR_PETBATTLE_QUEUE_SLOT_NO_SPECIES
* 5 = ERR_PETBATTLE_QUEUE_SLOT_CANT_BATTLE
* 6 = ERR_PETBATTLE_QUEUE_SLOT_REVOKED
* 7 = ERR_PETBATTLE_QUEUE_SLOT_DEAD
* 8 = ERR_PETBATTLE_QUEUE_SLOT_NO_PET
* default = ERR_PETBATTLE_QUEUE_JOIN_FAILED
8 = PET_BATTLE_QUEUE_PROPOSAL_ACCEPTED
9 = ERR_PETBATTLE_QUEUE_PROPOSAL_DECLINED
10 = ERR_PETBATTLE_QUEUE_OPPONENT_DECLINED
11 = ERR_PETBATTLE_QUEUE_PROPOSAL_TIMEOUT
12 = ERR_PETBATTLE_QUEUE_REMOVED
13 = ERR_PETBATTLE_QUEUE_REQUEUED_INTERNAL
14 = ERR_PETBATTLE_QUEUE_REQUEUED_REMOVED
18 = ?
19 = ?
20 = ERR_PETBATTLE_IN_BATTLE
21 = ERR_PETBATTLE_NOT_HERE

File: src/server/game/BattlePets/BattlePetMgr.h [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+
+    // wrong order
+    BATTLEPETRESULT_TOO_HIGH_LEVEL_TO_UNCAGE,
```
Comment (Warpten) [inline src/server/game/BattlePets/BattlePetMgr.h]: the fuck did you get these from

Comment (horn) [inline src/server/game/BattlePets/BattlePetMgr.h]: Shauren discovered the client enum in 6.1.2, but there were only names without the readable values

File: src/server/game/BattlePets/BattlePetMgr.h (hunk 17 above)
Comment (horn) [inline src/server/game/BattlePets/BattlePetMgr.h]: Thanks :)

**Comment by Warpten** [comment]
@horn status on this?

**Comment by spuresz** [comment]
coming soon

**Comment by horn** [comment]
Updated to 6.2.2 20444

**Pull request merged**